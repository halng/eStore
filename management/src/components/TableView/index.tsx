'use client'
import * as React from 'react'
import Paper from '@mui/material/Paper'
import Table from '@mui/material/Table'
import TableBody from '@mui/material/TableBody'
import TableCell from '@mui/material/TableCell'
import TableContainer from '@mui/material/TableContainer'
import TableHead from '@mui/material/TableHead'
import TablePagination from '@mui/material/TablePagination'
import TableRow from '@mui/material/TableRow'
import { ProductGroupTableData, ProductGroupTableHeader } from '@types'
import { Typography } from '@mui/material'

interface TableViewProps {
    tableHeader: ProductGroupTableHeader[] | null
    tableData: ProductGroupTableData[]
}

const TableView = ({ tableHeader, tableData }: TableViewProps) => {
    const [page, setPage] = React.useState(0)
    const [rowsPerPage, setRowsPerPage] = React.useState(10)

    const handleChangePage = (event: unknown, newPage: number) => {
        setPage(newPage)
    }

    const handleChangeRowsPerPage = (event: React.ChangeEvent<HTMLInputElement>) => {
        setRowsPerPage(+event.target.value)
        setPage(0)
    }

    if (!tableHeader || !tableData) {
        return (
            <Paper sx={{ width: '100%', textAlign: 'center', overflow: 'hidden' }}>
                <Typography variant='h5' component='h2'>
                    No Data!
                </Typography>
            </Paper>
        )
    }

    return (
        <Paper sx={{ width: '100%', overflow: 'hidden', paddingRight: '2rem' }}>
            <TableContainer sx={{ maxHeight: '40rem' }}>
                <Table stickyHeader aria-label='sticky table'>
                    <TableHead>
                        <TableRow>
                            {tableHeader.map((column) => {
                                if (column.isDisplay === false) return null
                                return (
                                    <TableCell
                                        key={column.id}
                                        align={column.align}
                                        style={{ minWidth: column.minWidth, fontWeight: 'bold', fontSize: '1rem' }}
                                    >
                                        {column.label}
                                    </TableCell>
                                )
                            })}
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {tableData.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage).map((row) => {
                            return (
                                <TableRow hover role='checkbox' tabIndex={-1} key={row.id}>
                                    {tableHeader.map((column) => {
                                        if (column.isDisplay === false) return null
                                        const value = row[column.id]
                                        return (
                                            <TableCell key={column.id} align={column.align}>
                                                {column.format && typeof value === 'number'
                                                    ? column.format(value)
                                                    : value}
                                            </TableCell>
                                        )
                                    })}
                                </TableRow>
                            )
                        })}
                    </TableBody>
                </Table>
            </TableContainer>
            <TablePagination
                rowsPerPageOptions={[10, 25, 100]}
                sx={{ pr: '2rem' }}
                component='div'
                count={tableData.length}
                rowsPerPage={rowsPerPage}
                page={page}
                onPageChange={handleChangePage}
                onRowsPerPageChange={handleChangeRowsPerPage}
            />
        </Paper>
    )
}

export default TableView
