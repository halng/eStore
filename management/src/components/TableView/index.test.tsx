import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import TableView from '.'
import { PRODUCT_GROUP_TABLE_HEADER } from '@constants'
import { mockProductGroupData } from '@mock'

describe('Test Rendering', () => {
    it('snapshot rendering', () => {
        const { asFragment } = render(<TableView tableData={[]} tableHeader={PRODUCT_GROUP_TABLE_HEADER} />)
        expect(asFragment()).toMatchSnapshot()
    })
    it('renders a heading', () => {
        render(<TableView tableData={mockProductGroupData} tableHeader={PRODUCT_GROUP_TABLE_HEADER} />)

        // check render table header
        const nameCol = screen.getByText('Name')
        const descriptionCol = screen.getByText('Description')
        const noCol = screen.getByText('No.')

        expect(nameCol).toBeInTheDocument()
        expect(descriptionCol).toBeInTheDocument()
        expect(noCol).toBeInTheDocument()

        // check render table data
        const nameCount = screen.getAllByText('Product Group', { exact: false })
        expect(nameCount.length).toBe(10)

        const combobox = screen.getByRole('combobox')
        expect(combobox).toBeInTheDocument()
    })
    it('renders a when missing data', () => {
        render(<TableView tableData={null} tableHeader={PRODUCT_GROUP_TABLE_HEADER} />)

        const noData = screen.getByText('No Data!')
        expect(noData).toBeInTheDocument()
    })
    it('renders a when empty data', () => {
        render(<TableView tableData={[]} tableHeader={PRODUCT_GROUP_TABLE_HEADER} />)

        // check render table header
        const nameCol = screen.getByText('Name')
        const descriptionCol = screen.getByText('Description')
        const noCol = screen.getByText('No.')

        expect(nameCol).toBeInTheDocument()
        expect(descriptionCol).toBeInTheDocument()
        expect(noCol).toBeInTheDocument()

        // check render table data
        const nameCount = screen.queryAllByText('Product Group')
        expect(nameCount).toHaveLength(0)
    })
})
