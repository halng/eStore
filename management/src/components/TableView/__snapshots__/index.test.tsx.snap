// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rendering snapshot rendering 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1cll31s-MuiPaper-root"
  >
    <div
      class="MuiTableContainer-root css-1qnqjkx-MuiTableContainer-root"
    >
      <table
        aria-label="sticky table"
        class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
              style="min-width: 10px; font-weight: bold; font-size: 1rem;"
            >
              No.
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
              style="min-width: 170px; font-weight: bold; font-size: 1rem;"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
              style="min-width: 170px; font-weight: bold; font-size: 1rem;"
            >
              Description
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
              style="min-width: 170px; font-weight: bold; font-size: 1rem;"
            >
              Last Update
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
              style="min-width: 70px; font-weight: bold; font-size: 1rem;"
            >
              Status
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
    <div
      class="MuiTablePagination-root css-14ed2tk-MuiTablePagination-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
      >
        <div
          class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
        />
        <p
          class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
          id=":r1:"
        >
          Rows per page:
        </p>
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-16c50h-MuiInputBase-root-MuiTablePagination-select"
        >
          <div
            aria-controls=":r2:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby=":r1: :r0:"
            class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
            id=":r0:"
            role="combobox"
            tabindex="0"
          >
            10
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="10"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
        >
          0â€“0 of 0
        </p>
        <div
          class="MuiTablePagination-actions"
        >
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to previous page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </button>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to next page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
